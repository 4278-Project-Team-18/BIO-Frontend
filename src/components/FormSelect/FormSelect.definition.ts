import type { SendInviteInputName } from "../SendInviteForm/SendInviteForm.definitions";
import type { Control } from "react-hook-form";
import type { NewClassInputName } from "../../modals/AddClassModal/AddClassModal.definitions";

export interface FormSelectProps {
  name: NewClassInputName | SendInviteInputName;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  control: Control<any, any>;
  label: string;
  options: FormSelectOption[];
  error?: string;
  optText?: string;
  defaultValue?: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  setValue?: any;
}

export interface FormSelectOption {
  value: string;
  label: string;
}
