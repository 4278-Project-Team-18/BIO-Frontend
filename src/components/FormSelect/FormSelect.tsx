import styles from "./FormSelect.module.css";
import { Controller } from "react-hook-form";
import { useEffect } from "react";
import type { FormSelectProps } from "./FormSelect.definition";

const FormSelect = ({
  name,
  control,
  label,
  error,
  optText,
  options,
  defaultValue,
  setValue,
}: FormSelectProps) => {
  useEffect(() => {
    if (defaultValue && setValue) {
      setValue(name, defaultValue);
    }
  }, []);

  return (
    <Controller
      name={name}
      control={control}
      render={({ field }) => (
        <div className={styles["form-select-container"]}>
          <div className={styles["form-select-inner-container"]}>
            {label && (
              <div
                className={
                  error
                    ? styles["form-select-label-no-opt"]
                    : styles["form-select-label-opt"]
                }
              >
                {label}
                {!error && optText && (
                  <span className={styles["form-select-opt-text"]}>
                    {" (" + optText + ")"}
                  </span>
                )}
                {error && (
                  <span className={styles["form-select-error"]}>
                    {" (" + error + ")"}
                  </span>
                )}
              </div>
            )}
            <select
              className={styles["form-select"]}
              onChange={field.onChange}
              onBlur={field.onBlur}
              name={field.name}
              ref={field.ref}
              defaultValue={defaultValue}
            >
              {options.map((option) => (
                <option
                  key={option.value}
                  value={option.value}
                  className={styles["form-select-option"]}
                >
                  {option.label}
                </option>
              ))}
            </select>
          </div>
        </div>
      )}
    />
  );
};

export default FormSelect;
