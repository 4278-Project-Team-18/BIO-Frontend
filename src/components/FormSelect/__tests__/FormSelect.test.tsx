import FormSelect from "../FormSelect";
import { render, renderHook } from "@testing-library/react";
import { useForm } from "react-hook-form";
import type { NewClassInputName } from "../../../modals/AddClassModal/AddClassModal.definitions";

describe("FormSelect Tests", () => {
  test("FormSelect should render correctly", () => {
    // Create a mock control
    const { result } = renderHook(() => useForm());
    const control = result.current.control;

    // Create mock props
    const mockProps = {
      name: "name" as NewClassInputName,
      label: "Test Label",
      options: [
        {
          value: "test1",
          label: "test1",
        },
        {
          value: "test2",
          label: "test2",
        },
      ],
      control: control,
      error: "Test Error",
      setValue: jest.fn(),
    };

    const component = render(<FormSelect {...mockProps} />);

    // Check if the component is rendered
    expect(component).toBeTruthy();

    // Check if the label is correct
    expect(component.getByText(mockProps.label)).toBeTruthy();

    // Check if the options are correct
    expect(component.getByText(mockProps.options[0].label)).toBeTruthy();
    expect(component.getByText(mockProps.options[1].label)).toBeTruthy();
  });

  test("FormSelect snapshot should render correctly", () => {
    // Create a mock control
    const { result } = renderHook(() => useForm());
    const control = result.current.control;

    // Create mock props
    const mockProps = {
      name: "name" as NewClassInputName,
      label: "Test Label",
      options: [
        {
          value: "test1",
          label: "test1",
        },
        {
          value: "test2",
          label: "test2",
        },
      ],
      control: control,
      error: "Test Error",
      setValue: jest.fn(),
    };

    const component = render(<FormSelect {...mockProps} />);

    expect(component).toMatchSnapshot();
  });
});
