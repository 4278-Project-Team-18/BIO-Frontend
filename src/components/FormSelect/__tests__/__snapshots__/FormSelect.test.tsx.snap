// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormSelect Tests FormSelect snapshot should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div />
    <div>
      <div
        class="form-select-container"
      >
        <div
          class="form-select-inner-container"
        >
          <div
            class="form-select-label-no-opt"
          >
            Test Label
            <span
              class="form-select-error"
            >
               (Test Error)
            </span>
          </div>
          <select
            class="form-select"
            name="name"
          >
            <option
              class="form-select-option"
              value=""
            />
            <option
              class="form-select-option"
              value="test1"
            >
              test1
            </option>
            <option
              class="form-select-option"
              value="test2"
            >
              test2
            </option>
          </select>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="form-select-container"
    >
      <div
        class="form-select-inner-container"
      >
        <div
          class="form-select-label-no-opt"
        >
          Test Label
          <span
            class="form-select-error"
          >
             (Test Error)
          </span>
        </div>
        <select
          class="form-select"
          name="name"
        >
          <option
            class="form-select-option"
            value=""
          />
          <option
            class="form-select-option"
            value="test1"
          >
            test1
          </option>
          <option
            class="form-select-option"
            value="test2"
          >
            test2
          </option>
        </select>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
